File: cses/cses_Grid_Path_Description.py
paths = []
directions = {
    'L': (-1,  0),
    'R': ( 1,  0),
    'U': ( 0, -1),
    'D': ( 0,  1),
}

valid = "LRUD"
n=7
lm=48

templ = '?'*lm
board = [
    [0,1,1,1,1,1,1,1],
    [1,1,1,1,1,1,1,1],
    [1,1,1,1,1,1,1,1],
    [1,1,1,1,1,1,1,1],
    [1,1,1,1,1,1,1,1],
    [1,1,1,1,1,1,1,1],
    [1,1,1,1,1,1,1,1],
    [1,1,1,1,1,1,1,1],
]

def move(moves=[' ' for i in range(48)],cx=0,cy=0,left=lm):
    if left==0:
        if cx==0 and cy==7:
            paths.append(moves)       
        return
    
    touse = valid if templ[lm-left] == '?' else templ[lm-left]
    for turn in touse:
        h = directions[turn][0]
        v = directions[turn][1]
        nx = cx+h
        ny = cy+v
        if nx >= 0 and nx < n and ny >= 0 and ny < n and board[nx][ny]==1:
            board[nx][ny] = 0
            moves[lm-left] = turn
            move(moves,nx,ny,left-1)
            moves[lm-left] = ' '
            board[nx][ny] = 1
    return

t = input()
move()
print(len(paths))

File: cses/cses_Grid_Path_Description.cpp
#include<bits/stdc++.h>
using namespace std;

int tot=0;
int lm=48;

int templ[48]; // -1 is ?
bool board[9][9]; //0 means empty //-1 means dont touch

int dx[4] = {-1,1,0,0};
int dy[4] = {0,0,-1,1};
char d[4] = {'L','R','U','D'};

void move(int cy=1, int cx=1, int left=lm){

    if((board[cy-1][cx] != 0 and board[cy+1][cx] != 0 and board[cy][cx-1]==0 and board[cy][cx+1]==0)
    or (board[cy][cx-1] != 0 and board[cy][cx+1] != 0 and board[cy-1][cx]==0 and board[cy+1][cx]==0)){
        return;
    }
    
    if(cy==7 and cx==1){
        if(left==0) tot++;
        return;
    }

    for(int i=0; i<4; i++){
        if(templ[lm-left] != -1 and templ[lm-left] != i) continue;
        int ny = cy+dy[i];
        int nx = cx+dx[i];
        if(board[ny][nx]==0){
            board[ny][nx] = 1;
            move(ny,nx,left-1);
            board[ny][nx] = 0;
        }
    }
}

int main(){
    ios::sync_with_stdio(false); cin.tie(nullptr);

    for(int i=0; i<48; i++) templ[i] = -1;
    for(int i=0; i<9; i++){
        board[0][i] = -1;
        board[8][i] = -1;
    }
    for(int i=1; i<8; i++){
        board[i][0] = -1;
        board[i][8] = -1;
    }
    
    board[1][1] = 1;
    string s; cin >> s;
    for(int i=0; i<48; i++){
        if(s[i] == 'L') templ[i] = 0;
        else if(s[i] == 'R') templ[i] = 1;
        else if(s[i] == 'U') templ[i] = 2;
        else if(s[i] == 'D') templ[i] = 3;
    }
    
    move();
    cout << tot << "\n";
}